syncs/syncs.go

```diff
@@ -304,3 +304,39 @@ func (wg *WaitGroup) Go(f func()) {
 		f()
 	}()
 }
+
+// TODO(https://go.dev/issue/63941): LockFunc, LockValue, and LockValues
+// are helper functions proposed upstream. Their naming and signature
+// are based on the existing OnceFunc, OnceValue, and OnceValues
+// helper functions already in the [sync] package.
+
+// LockFunc runs f while holding the lock.
+func LockFunc(lock sync.Locker, f func()) {
+	lock.Lock()
+	defer lock.Unlock()
+	f()
+}
+
+// LockValue runs f while holding the lock and returns the argument.
+func LockValue[T any](lock sync.Locker, f func() T) T {
+	lock.Lock()
+	defer lock.Unlock()
+	return f()
+}
+
+// LockValues runs f while holding the lock and returns the arguments.
+func LockValues[T1, T2 any](lock sync.Locker, f func() (T1, T2)) (T1, T2) {
+	lock.Lock()
+	defer lock.Unlock()
+	return f()
+}
+
+// Mutex is identical to [sync.Mutex], but with additional methods.
+type Mutex struct{ sync.Mutex }
+
+// Do runs f while holding the lock.
+func (m *Mutex) Do(f func()) {
+	m.Lock()
+	defer m.Unlock()
+	f()
+}
```

Comment:
@raggi has aspirations of getting static analysis doing lock checking in the tree. I wonder what this would do both to that and to the readability/consistency of mutex usage in the tree.

I don't want to end up in a place where we have a bunch of different locking styles.

Reply:
I'd started with gvisors static lock checks, sadly they don't yet support various patterns we use a lot. This pattern is helpful there as it disallows one of the harder cases.

Reply:
I'm not exactly sure what @raggi had in mind, but I would argue that this would assist future static analysis.

One challenge in static analysis is correlating the set of variables that are actually guarded by a lock to a particular lock. Today, the pattern of doing:
```go
mu.Lock()
defer mu.Unlock()

... // critical section

... // uncritical section accidentally covered by mu since defers execute late
```

Consequently, we might have more variables covered by a lock than is necessary.

The initial motivation of this was to make the critical sections more clear in code.

Reply:
https://github.com/tailscale/tailscale/actions/workflows/checklocks.yml was running for a while in a non-blocking mode. Unfortunately it has problems with bounds as you describe, but most particularly with patterns like:

```
m.Lock()
defer m.Unlock()

go func() { m.Lock(); defer m.Unlock(); ... }
```

Which show up in a good number of places in our code, in either local or non-local form.

We could do with cleaning those cases up too, for other reasons, but as it stands this blocks using the tool.